File: backend/src/lib/mealPeriod.ts
/**
 * Meal period rules (Ghana time — UTC+0, server runs UTC):
 *   Breakfast:     08:00 – 10:00
 *   Lunch & Dinner: 10:00 – 20:00
 *   Closed:        20:00 – 08:00
 *
 * The server stores all times in UTC. Ghana is GMT+0 (no DST),
 * so UTC hour == Ghana hour.
 */

export type MealPeriod = 'BREAKFAST' | 'LUNCH_DINNER' | 'CLOSED';

export function getMealPeriod(date: Date = new Date()): MealPeriod {
  const hour = date.getUTCHours();
  if (hour >= 8 && hour < 10) return 'BREAKFAST';
  if (hour >= 10 && hour < 20) return 'LUNCH_DINNER';
  return 'CLOSED';
}

export function getMealPeriodLabel(period: MealPeriod): string {
  switch (period) {
    case 'BREAKFAST': return 'Breakfast';
    case 'LUNCH_DINNER': return 'Lunch & Dinner';
    case 'CLOSED': return 'Closed';
  }
}

/** Returns the end of the current operating window (UTC), or null if closed. */
export function currentPeriodEnd(date: Date = new Date()): Date | null {
  const period = getMealPeriod(date);
  if (period === 'CLOSED') return null;
  const end = new Date(date);
  end.setUTCMinutes(0, 0, 0);
  end.setUTCHours(period === 'BREAKFAST' ? 10 : 20);
  return end;
}

/** Validates that a pickupWindow falls within an active meal period. */
export function isValidPickupWindow(pickupWindow: Date): boolean {
  return getMealPeriod(pickupWindow) !== 'CLOSED';
}

File: mobile_app/lib/utils/meal_period.dart
enum MealPeriod { breakfast, lunchDinner, closed }

class MealPeriodHelper {
  /// Returns the current meal period based on UTC hour (Ghana is GMT+0).
  static MealPeriod current() {
    final hour = DateTime.now().toUtc().hour;
    if (hour >= 8 && hour < 10) return MealPeriod.breakfast;
    if (hour >= 10 && hour < 20) return MealPeriod.lunchDinner;
    return MealPeriod.closed;
  }

  static String label(MealPeriod p) {
    switch (p) {
      case MealPeriod.breakfast: return 'Breakfast';
      case MealPeriod.lunchDinner: return 'Lunch & Dinner';
      case MealPeriod.closed: return 'Closed';
    }
  }

  /// Returns a pickupWindow 30 minutes from now, capped within the current period end.
  /// Returns null if currently closed.
  static DateTime? pickupWindow() {
    final now = DateTime.now().toUtc();
    final period = current();
    if (period == MealPeriod.closed) return null;

    final endHour = period == MealPeriod.breakfast ? 10 : 20;
    final periodEnd = DateTime.utc(now.year, now.month, now.day, endHour);

    final desired = now.add(const Duration(minutes: 30));
    // Clamp to 5 minutes before period end so pickup is always valid
    final latest = periodEnd.subtract(const Duration(minutes: 5));
    return desired.isBefore(latest) ? desired : latest;
  }
}
